dvc/utils/threadpool.py

```diff
@@ -0,0 +1,32 @@
+from concurrent import futures
+from itertools import islice
+from typing import Any, Callable, Iterable, Iterator, Set, TypeVar
+
+_T = TypeVar("_T")
+
+
+class ThreadPoolExecutor(futures.ThreadPoolExecutor):
+    @property
+    def max_workers(self) -> int:
+        return self._max_workers
+
+    def imap_unordered(
+        self, fn: Callable[..., _T], *iterables: Iterable[Any]
+    ) -> Iterator[_T]:
+        """Lazier version of map that does not preserve ordering of results.
+
+        It does not create all the futures at once to reduce memory usage.
+        """
+
+        def create_taskset(n: int) -> Set[futures.Future]:
+            return {self.submit(fn, *args) for args in islice(it, n)}
+
+        it = zip(*iterables)
+        tasks = create_taskset(self.max_workers * 5)
+        while tasks:
+            done, tasks = futures.wait(
+                tasks, return_when=futures.FIRST_COMPLETED
+            )
+            for fut in done:
+                yield fut.result()
+            tasks.update(create_taskset(len(done)))
```

Comment:
Btw, there was a plan to try using async here e.g. with fsspec's put/get methods. Maybe it would be reasonable to do that instead now?

Reply:
Could you please elaborate? I am not sure I understand.

Reply:
So we use thread pool executor right now to parallelize uploading/downloading of objects, but fsspec provides `put/get` methods that accept batches and  async filesystems do that with async, effectively managing the coroutines, instead of us manually using threads.

Reply:
So what you are suggesting is to have something like TransferManager that handles multithreaded and async transfer as well right?

Reply:
@skshetry Ah, no, I meant that we possibly won't need transfer manager at all, and could just `fs.put(obj_list)` and let it do its thing. So we would be effectively delegating transfer management to particular filesystem's put/get methods.

Reply:
@skshetry Btw, happy to jump on a call to talk about it a bit more.

Reply:
I think we should wait for fsspec changes to be done, this will likely require us to have colored APIs too (`transfer_async`, `odb.add_async` etc.).  And, there's a problem of running fsspec's sync FS in async mode too. 

As this change is very minor restructuring, I'd prefer to merge this as-is.